### ISSUE TITLE
Docker configuration ignored with "port is missing"

### ISSUE THREAD

**Issue by xsrf** [containous/traefik#5732]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I tried to run traefik in a docker swarm and wanted to expose the dashboard as shown in the docs here: https://docs.traefik.io/operations/dashboard/#secure-mode

### What did you expect to see?

The dashboard on `http://127.0.0.11/dashboard/`

### What did you see instead?

A 404 from traefik

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.2
Codename:     montdor
Go version:   go1.13.1
Built:        2019-10-09T19:26:05Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```yml
version: '3.7'
services:
  traefik:
    image: traefik:latest
    ports:
      - 80:80
      - 443:443
    deploy:
      replicas: 1
      placement:
        constraints:
          - node.role == manager
      labels:
        - traefik.enable=true
        - traefik.http.routers.traefik_http.rule=Host(`127.0.0.11`)
        - traefik.http.routers.traefik_http.service=api@internal
        - traefik.http.routers.traefik_http.entrypoints=http
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command: >
      --providers.docker
      --providers.docker.exposedbydefault=false
      --providers.docker.swarmmode=true
      --entryPoints.http.address=":80"
      --entryPoints.https.address=":443"
      --accesslog
      --log.level=INFO
      --api=true
      --api.dashboard=true
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
level=error msg="port is missing" providerName=docker container=traefik-traefik-1jvz83nonxx53q9eel9r4cwlg
```

Traefik seems to ignore configurations of a docker service if they are missing a service definition setting a loadbalancer port.
Since `api@internal` is an internal service which is already defined and has no port, I didn't define one.
However, if I add a **dummy service** with a port, **the Dashboard works fine**:

```yml
...
      labels:
        - traefik.enable=true
        - traefik.http.services.dummyService.loadbalancer.server.port=1337
        - traefik.http.routers.traefik_http.rule=Host(`127.0.0.11`)
        - traefik.http.routers.traefik_http.service=api@internal
        - traefik.http.routers.traefik_http.entrypoints=http
...
```

I did some more experiments:

```yml
# Route and Services work fine:
  whoami:
    image: containous/whoami
    deploy:
      replicas: 2
      labels:
        - traefik.enable=true
        - traefik.http.services.whoami.loadbalancer.server.port=80
        - traefik.http.routers.whoami_http.rule=Host(`127.0.0.12`)
        - traefik.http.routers.whoami_http.service=whoami
        - traefik.http.routers.whoami_http.entrypoints=http

# Route for whoami2 is ignored because of missing port:
  whoami2:
    image: containous/whoami
    deploy:
      replicas: 1
      labels:
        - traefik.enable=true
        - traefik.http.routers.whoami2_http.rule=Host(`127.0.0.13`)
        - traefik.http.routers.whoami2_http.service=whoami
        - traefik.http.routers.whoami2_http.entrypoints=http
      
# Route for whoami3 is ignored because of missing port:
  whoami3:
    image: containous/whoami
    deploy:
      replicas: 1
      labels:
        - traefik.enable=true
        - traefik.http.routers.whoami3_http.rule=Host(`127.0.0.14`)

# Route for whoami4 is parsed and served by whoami4:
  whoami4:
    image: containous/whoami
    deploy:
      replicas: 1
      labels:
        - traefik.enable=true
        - traefik.http.services.whoami4.loadbalancer.server.port=80
        - traefik.http.routers.whoami4_http.rule=Host(`127.0.0.15`)

# Route for whoami5 is parsed and served by whoami:
  whoami5:
    image: containous/whoami
    deploy:
      replicas: 1
      labels:
        - traefik.enable=true
        - traefik.http.services.whoami5.loadbalancer.server.port=1337
        - traefik.http.routers.whoami5_http.rule=Host(`127.0.0.16`)
        - traefik.http.routers.whoami5_http.service=whoami
```
Errors here:
```
level=error msg="port is missing" providerName=docker container=traefik-whoami2-lvaeohb9ycx9g8vqs5bhhep5b
level=error msg="port is missing" providerName=docker container=traefik-whoami3-mpr8w2rl8lx21b13vt3c3nm0e
```

It looks like traefik requires a `traefik.http.services.*.loadbalancer.server.port` label on every service, or it will discard the configuration. Name, existence and port of the service are completely irrelevant.

Traefik should not require a loadbalancer port definition for each docker service, because a router can also reference an internal service or a service defined elsewhere and thus defining a service/loadbalancer is pointless.

**Comment by ldez** [comment]
Hello,

In Swarm mode, the port cannot be automatically detect, so the port must be manually set.

> Mandatory for Docker Swarm.

https://docs.traefik.io/v2.0/routing/providers/docker/#services

**Comment by xsrf** [comment]
I understand that, but I still think this is an error.
It explains why, in my example, `whoami3` is ignored. `whoami3` has no service definition, so the service is automatically generated. As you stated, this does not work in swarm.

However, the fact that `whoami2` is ignored, is an error. `whoami2` refers to a service with a defined loadbalancer port. The same applies to my initial problem with `api@internal`.
There shouldn't be a pointless dummy port being required, like what makes `whoami5` work.

**Comment by luis100** [comment]
Same here, there seems to be no other way to enable traefik dashboard secure mode in Docker Swarm without creating a dummy service.

**Comment by AndrewSav** [comment]
>There shouldn't be a pointless dummy port being required, like what makes whoami5 work.

@ldez I think @xsrf has a good point here. There are clearly cases when this port is not needed or used, yet without it the provider configuration for the affected container is not being sent.

I think you are aware of that given this post:

https://community.containo.us/t/dashboard-gives-me-an-error-bad-gateway/1883/2

In particular it says:

```yaml
- "traefik.http.services.traefik.loadbalancer.server.port=888" # required by swarm but not used.
```

I'm not sure I understand why it is required by _swarm_. It seems to be required by traefik, but why is it required if it is not used?

**Comment by AndrewSav** [comment]
Perhaps [buildServiceConfiguration](https://github.com/containous/traefik/blob/v2.0.4/pkg/provider/docker/config.go#L99) should be modified to inspect `configuration.Routers[*].Service`. If each of the routers already have a service set, then `buildServiceConfiguration` does not need to create a default one?

**Comment by ldez** [comment]
To understand the behavior:

when no explicit router or service: (ie: no labels)
- a router and a service are automatically created

when a router is define but no service: (ie: `traefik.http.routers.xxx.yyy` label[s])
- a service is automatically created

The automatic creation of the service is not related to the router (ie: not dependent of `traefik.http.routers.<router_name>.service`)

**Comment by AndrewSav** [comment]
Yep, this is understood. All the above I said still applies -may be it's time to change status quo. 

The main issue here, is that when you have a router that points to an existing service (`traefik.http.routers.<router_name>.service`) _and_ there is no service explicitly defined, chances are you do not need to automatically create a service.

Trying to create a service that was not requested, failing and refusing to apply the rest of labels in swarm mode is not nescessary, detremental and surprising. I'm proposing to improve this.

Do you agree that there is a problem here, that would benefit from a fix?

If it is desirable to attempt to create a service automatically even if there is no router without an explicit service (so the service is unlikely to be used), then the improvement should be, when failing to create a service not to fail the rest of configuration for the docker container/service, I currently cannot see downsides of this.

**Comment by AndrewSav** [comment]
Another related question for me to understand the issue better: Why do we need to fail [here](https://github.com/containous/traefik/blob/v2.0.4/pkg/provider/docker/config.go#L197) at all? Can we just not create service (Log it, retrun without a error)? What will break if we do? Router without services will expect at least one service, but we can change that too, cannot we, and fail _there_, when we know that the service would actually be used if it existed.

**Comment by ldez** [comment]
> Why do we need to fail here at all?

yes

> Can we just not create service (Log it, return without a error)?

no

> What will break if we do?

yes

> Do you @ldez agree that there is a problem here, that would benefit from a fix?

it's a bit more complex: it's the expected behavior for common case but in some case it can be problem.

**Comment by AndrewSav** [comment]
Usually why/what questions reqruire a bit more than "yes" or "no" ;)))__

>This line is not the root of the behavior

Oh, you know the code base and I don't - you wrote it, I don't doubt you. I'm just seeing an issue that is being brought up again and again on the forums, and I'm trying being helpful to find a solution.

>it's a bit more complex: it's the expected behavior for the common cases but in some cases it can be a problem.

It seems to me that it is possible to fix the problem and not break the common cases. I tried to explain that above, but perhaps not adequatly.

It feels like you have it under control and don't have time to discuss this in more details, so I'll leave this up to you. Let me know if you want discuss / my help.

**Comment by msonowal** [comment]
@xsrf 
Any solutions yet,
@traefiker  Please someone should look into this as critical issue

**Comment by AndrewSav** [comment]
@msonowal well, the issue is understood and workaround is known, as per discussion above. I would not call it critical, in fact it's pretty minor. Traefik team is well aware of this issue as evidenced by post from ldez, traefik team member.

**Comment by msonowal** [comment]
@AndrewSav 
I too understood about that, but it should be mentioned in the docs as peoples which are new or upgrading from 1.7 need to aware of this

**Comment by AndrewSav** [comment]
@msonowal it is mentioned in the docs. Look at <https://docs.traefik.io/routing/providers/docker/#services>. Under `traefik.http.services.<service_name>.loadbalancer.server.port` subsection it says: `Mandatory for Docker Swarm.`

**Comment by msonowal** [comment]
@AndrewSav 
I saw that but in case for deploying traefik as in the cluster its not appropriate that we define the port for that label, so this is a confusing for people

**Comment by dduportal** [comment]
@msonowal What do you mean? Not sure to understand if you are talking a specific part of the documentation or about the feature requirement itself?

**Comment by cookiejest** [comment]
just spent like 3 hours on this. Not clear from the docs that this flag was needed anywhere in the docs.. nearly made me not use traefik!

**Comment by msonowal** [comment]
@dduportal 
I am referring to the thing when deploying traefik as a service and how to use the traefik dashboard as it is a internal special kind of service so people will not simply set the load balancer port for the traefik service itself thats why it is confusing peple

**Comment by cookiejest** [comment]
@AndrewSav  just a little note in the same place to say it needs to be on the treafik service itself as well would be helpful if nothing else :)

**Comment by dduportal** [comment]
Hi @msonowal @cookiejest @xsrf , we hear you, and a first step has been done when merging #5795 by adding a Docker Swarm example, separately than "Docker", for both Dashboard (https://docs.traefik.io/v2.0/operations/dashboard/#secure-mode) and API (https://docs.traefik.io/v2.0/operations/api/#configuration).

**Comment by remotezygote** [comment]
There's more to this. You also get into this situation if you apply a middleware from a file to a swarm service. Example:

`/etc/traefik/dynamic/stripServerHeader.yml`:
```
http:
  middlewares:
    stripServerHeader:
      headers:
        customResponseHeaders:
          Server: ""
```

`/etc/traefik/traefik.yml`:
```
providers:
  docker:
    endpoint: 'unix:///var/run/docker.sock'
    swarmMode: 'true'
    exposedByDefault: 'false'
    network: 'traefik'
  file:
    directory: /etc/traefik/dynamic/

entryPoints:
  http:
    address: ':8000'
  ping:
    address: ':8083'
  prometheus:
    address: ':8084'

ping:
  entrypoint: 'ping'

metrics:
  prometheus:
    entrypoint: 'prometheus'

accessLog: {}

retry:
  attempts: '0'

global:
  sendAnonymousUsage: false
```

`docker-compose.yml`:
```
services:
  myservice:
    deploy:
      labels:
        traefik.enable: 'true'
        traefik.http.services.myservice.loadbalancer.server.port: '3000'
        traefik.http.middlewares.myservice-ratelimit.ratelimit.average: '3000'
        traefik.http.middlewares.myservice-ratelimit.ratelimit.burst: '6000'
        traefik.http.routers.myservice.rule: PathPrefix(`/`)
        traefik.http.routers.myservice.middlewares: myservice-ratelimit@docker, stripServerHeader@file
        traefik.http.routers.myservice.entrypoints: http
```

With this configuration, every service I apply the `stripServerHeader@file` middleware to gets the `port is missing` error (even though they all have ports). Traefik does not seem to handle any of these services at all, like it completely ignored them.

If I remove the `stripServerHeader@file` middleware, I get no errors, but do not get the middleware functionality.

**Comment by remotezygote** [comment]
That's not in any way remotely off topic.

**Comment by dduportal** [comment]
@remotezygote the topic already stated that defining the port is mandatory in Swarm. If you define it, you should not have any problem with middleware. We consider any additional comment as offline if not adding anything to the discussion, in particular when there is PR open to fix the documentation. Don’t worry, the goal is not to ignore you: there is the community forum if you want some help to fix your problem of course! We just keep the issue tracker for bugs and feature requests, which is not the topic here :)

**Comment by remotezygote** [comment]
Please read my post again. I have presented a new case where this same error occurs _when the port is set_.

This is not solved.

**Comment by AndrewSav** [comment]
@remotezygote Could you please double check your configuration. It's possible that you get this error due a different part of configuration you are not showing. If you are confident, that you are correct, please provide a short reproducible example that others can run to see your issue.

**Comment by remotezygote** [comment]
I posted the entire configuration. Try it.

**Comment by dduportal** [comment]
@remotezygote I cannot reproduce, even though your configuration is partial: how is swarm deployed? And Traefik? And your service? Does it have data volumes? External networks? something else?

Can you open a topic on the community forum? If it appears to be a reproducible bug, we'll open an issue for this. But until then, as we cannot reproduce, this is not the place for getting help here.

**Comment by remotezygote** [comment]
I added a comment here: https://community.containo.us/t/global-ip-whitelisting/1816

You need a solution for globally-applied middlewares, and fast.

**Comment by ldez** [comment]
Closed by #5874

**Comment by xsrf** [comment]
Sorry, but I don't consider this done. Just documenting that you have to add a silly dummy value to your configuration isn't really solving the issue. It's a workaround at best.
It's also only addressing the issue with the Dashboard/API.
It is not addressing the issues with my example config where one docker service is actually referencing an others docker service traefik service.

**Comment by Pylipala** [comment]
Hey, I am still facing the same issue, my docker-compose.yml is as following:
```
version: '3'

services:
  reverse-proxy:
    # The official v2.0 Traefik docker image
    image: traefik
    # Enables the web UI and tells Traefik to listen to docker
    command: >
      --providers.docker
      --providers.docker.exposedbydefault=true
      --providers.docker.swarmmode=true
      --entryPoints.http.address=":80"
      --entryPoints.https.address=":443"
      --accesslog
      --log.level=TRACE
      --api=true
      --api.dashboard=true
      --api.insecure=true
    ports:
      # The HTTP port
      - "180:80"
      # The Web UI (enabled by --api.insecure=true)
      - "18080:8080"
    volumes:
      # So that Traefik can listen to the Docker events
      - /var/run/docker.sock:/var/run/docker.sock

  whoami:
    # A container that exposes an API to show its IP address
    image: containous/whoami
    labels:
      - "traefik.http.routers.whoami.rule=Host(`whoami.docker.localhost`)"
      - "traefik.http.services.whoami.loadbalancer.server.port=80"
```
But it prints in `docker logs 1eb3ef123b8b`
```
time="2019-12-08T15:02:01Z" level=error msg="port is missing" providerName=docker container=local-whoami-lgo9ik9s4sxqkaw0b38os7fgr
time="2019-12-08T15:02:01Z" level=error msg="port is missing" providerName=docker container=local-reverse-proxy-smsb8kyuqe5y0pwazgfazkwj8
```
and `curl http://whoami.docker.localhost:180` prints `404 page not found`

**Comment by ldez** [comment]
@Pylipala the labels must be placed into the section `deploy` with Docker Swarm.
https://docs.traefik.io/v2.1/providers/docker/#routing-configuration-with-labels_1

**Comment by felixsanz** [comment]
@xsrf @AndrewSav why not add an if condition that checks if the service is internal so no port is required? i've wasted some hours on this too, makes no sense

**Comment by ldez** [comment]
WIth Docker, the port is autodetected but With Swarm is impossible to know the port (related to swarm networking).

**Comment by felixsanz** [comment]
@ldez Why is needed to know the port of `api@internal`? I don't get it

**Comment by ldez** [comment]
It's a issue with swarm and the internal service, I already explain that: https://github.com/containous/traefik/issues/5732#issuecomment-549096489

**Comment by EthraZa** [comment]
After a full day trying to get Traefik, even after finding this conversation about the dummy line, I still can't get it to work. Traefik makes Haproxy look freaking easy to setup.

**Comment by kriscdev** [comment]
> @ldez @AndrewSav why not add an if condition that checks if the service is internal so no port is required? i've wasted some hours on this too, makes no sense

Echoing the above -- while I am but a shade-tree home lab hero, also a fan of logic and principles, and the current state violates both. There is no defensible position for the workaround, other than lack of transparency/an oversight/laziness. While I do this for a hobby, I'd be absolutely livid if my livelihood rested on such a faulty premise and fallacious logic. Read: I would quickly abandon Traefik altogether. There's no defending this choice, as conclusory as that assertion may be, what's on this thread is sound logic and should not be ignored.

**Comment by MicahZoltu** [comment]
> To understand the behavior:
> 
> when no explicit router or service: (ie: no labels)
> 
>     * a router and a service are automatically created

@ldez is it possible to disable this behavior?  I have a lot of services that are purely internal and definitely should *not* be routed to by traefik.  I have a relatively small number of public facing services that *should* be routed to by traefik.  I would like the default behavior, when a service has no traefik labels, to be that traefik ignores the service and it only pays attention to the service (and thus complains about it being misconfigured) when I explicitly add some traefik configuration.

**Comment by ldez** [comment]
@MicahZoltu Take a look to https://doc.traefik.io/traefik/v2.4/providers/docker/#exposedbydefault

**Comment by MicahZoltu** [comment]
> @MicahZoltu Take a look to https://doc.traefik.io/traefik/v2.4/providers/docker/#exposedbydefault

Thanks!  Maybe I'm misunderstanding people above, but I feel like this should be the answer to this question.  😄

**Comment by binaryDiv** [comment]
Just wanted to note that this issue is not exclusive to Swarm.

I just changed my Traefik container (normal Docker without Swarm) to use host networking, thus removing the "port" definition in docker-compose.yml. Everything worked fine except for the Traefik Dashboard, same issue as above: "port is missing". Adding this dummy service "solved" the problem:

```
labels:
  traefik.enable: true
  traefik.http.routers.api.rule: Host(`traefik.example.com`)
  traefik.http.routers.api.service: api@internal
  traefik.http.services.dummy-svc.loadbalancer.server.port: 9999
```

This isn't really a solution though, it's a silly workaround that should not be necessary.

I see why Traefik throws an error when it can't detect the port to autocreate a service. But in this case, there is no need to create a service at all, because an existing internal service is used.

I think the solution is actually pretty simple: If a container defines a router with an `@internal` service, don't try to autocreate an additional service (unless it's explicitly defined using `traefik.http.services.[...]` labels). Or maybe do try to autocreate it (if that still makes sense in some way), but don't ignore the *router* labels when service autocreation fails, since those router labels are independent from that autocreated service.

**Comment by wookiesh** [comment]
Just stumbled upon this issue... 
Was working, then not, not changing anything, is it really not yet solved ?

**Comment by mfloresVicomtech** [comment]
Hi, just to note that this also applies to AWS ECS provider.

**Comment by wookiesh** [comment]
Thanks for the support, the workaround is indeed working, but I'm trying to
push the usage of traefik in production and I have to admit that this does
not help me to sell it to the team.

**Comment by dseven** [comment]
Totally agree with @binaryDiv here. I ran into this in the same way - traefik container using host networking. At the very least, the kludgearound should be documented for "Docker" as well as "Docker (swarm)". It would have saved me another headache...

**Comment by SayidHosseini** [comment]
@binaryDiv Thank you that works.

Spent a lot of time finding this issue and work around.
I hope this gets documented soon enough😢

**Comment by uniuuu** [comment]
> @MicahZoltu Take a look to
> 
>     * https://doc.traefik.io/traefik/v2.4/providers/docker/#exposedbydefault
> 
>     * https://doc.traefik.io/traefik/v2.4/providers/overview/#restrict-the-scope-of-service-discovery

Setting next: ```exposedByDefault = false```  resolves the issue.

**Comment by rachelf42** [comment]
i am also having this issue, not using swarm, and exposedByDefault was already false
dummy port solution worked

**Comment by PixelHir** [comment]
how is this still an open issue? running traefik on host networking, trying to access the dashboard, won't work without the dummy port

**Comment by SR-G** [comment]
Have just lost 2 hours on this...
How can this be possible...
The documentation does NOT contain anymore the dummy service line, as far as i can tell (nothing here for example : https://doc.traefik.io/traefik/operations/api/#configuration)

**Comment by tfny** [comment]
Hey @SR-G and @cnmicha, 

I see what happened, the information is in that section, you just have to switch tabs above the code block to select Docker Swarm. 
![Screen Shot 2022-11-08 at 8 53 09 AM](https://user-images.githubusercontent.com/95721958/200597110-593786e2-e57b-4ea8-b900-a13f4a9934a6.png)

I am also going to link the documentation for port detection incase that helps you out. https://doc.traefik.io/traefik/providers/docker/#port-detection_1

**Comment by everyx** [comment]
It is really puzzling to define the port when there is no need to define the port. Can Traefik ignore the port check when we use the service is already defined in the dynamic file?

**Comment by gumbo2k** [comment]
Yeah. This one caused me some headache as well.

I have a slightly different use case, I don't want to expose an `@internal` service, but I want to define some common `middlewares` that I want to re-use on several of my 50+ services.

I work on docker swarm and those 50+ services currently are loadbalanced by an nginx container.

Most services fall into 2 categories and could use some of the same middlewares. 
I want to define those middlewares via labels on the `traefik` container itself, so that I can re-use them on the services and I can manage them in *one* place, instead of  duplicating it on *every single service*.

Currently this would look like this Example:
```
            labels:
                traefik.enable: "true"
                # HERE IS THE DUMMY STUFF
                traefik.http.services.dummy-service.loadbalancer.server.port: 9999
                traefik.http.routers.dummy-route.entrypoints: web
                traefik.http.routers.dummy-route.service: noop@internal

                # HERE is the stuff that I REALLY want to define.
                traefik.http.middlewares.slowrate.ratelimit.average: 5
                traefik.http.middlewares.slowrate.ratelimit.burst: 10

                traefik.http.middlewares.fastrate.ratelimit.average: 100
                traefik.http.middlewares.fastrate.ratelimit.burst: 500

                traefik.http.middlewares.source-ip-is-private.ipwhitelist.sourcerange: 10.0.0.0/8,172.16.0.0/12,192.168.0.0/16,127.0.0.1/8
... and so on ... 

```

I really want to migrate to traefik, so that I don't need to rebuild the nginx container every time we add a new service, some rate limit changes or some some api endpoint is added ... but currently I have to employ **two** workarounds to get this up and running... 

* I have to define a service
* I have to define a route

Both, the route and the service, simply speaking should not be needed to declare middlewares that are not even used by them.

There might be workarounds, by declaring middlewares in files, but seriously? I want to switch to traefik, _to get rid of config files_ .. be that as part of a custom built image, or as a separate artifact that needs to be deployed.

**Comment by Froosh** [comment]
More fun: this is not limited to docker swarm mode, it also happens with network_mode: host

**Comment by dseven** [comment]
> More fun: this is not limited to docker swarm mode, it also happens with network_mode: host

.... as noted by @binaryDiv two years ago: https://github.com/traefik/traefik/issues/5732#issuecomment-894362338

**Comment by cookiejest** [comment]
i appreciated the technical challenge that traefik provided. but tbh the juice wasnt worth the squeeze and we ended up just using AWS load balancer to solve our problems and a much more simple approach

**Comment by rtribotte** [comment]
Hello,

I took a look at the Docker provider implementation, and hence the Docker Swarm provider, as the same code to build the configuration is shared between the two.

Considering the use case initially exposed by @xsrf, which is to be able to provide the router's configuration exposing internal services (or other services than the container one), while being in SwarmMode and not having to provide the "dummy" port label, I wanted to share the following proposal to make it possible.

Only in Swarm, and when building the service configuration from a container (service) that has no service configuration labels, we could be resilient from the error that will happen because the service port is not defined.

This issue lies in the fact that the Docker provider (Swarm as well), is building the services configuration before building the routers one.
And, for one container, if an error is raised during the parsing of the services configuration, the router's configuration will be skipped.
Being resilient from the error, in that very situation, would allow us to parse anyway the router's configuration, and support the use case exposed in this issue.

The other consequence, when the container router's configuration is actually using the container's service, will be that the error is no longer raised during the parsing of the configuration, but when building the actual routers and services from it.
In other words, a missing service port label when using Swarm would not be silent but would raise a different error, less meaningful since it would only state that the router's service is missing.
To find out what would have happened with that service, we must still warn in the logs that the service was not built because the port was missing.

**Comment by chris-pinbn** [comment]
Going to add my 2c: I have a situation similar in that I would love the ability to define a "default port" for services, so that a route that references a service (in docker swarm) where no port is defined could instead use a "default port" defined at the top traefik level (e.g. website hosting, all images run on port 80, so rather than having to add a line for all 100+ sites the cluster runs, I could say "default is 80 unless otherwise specified"). Would that not resolve this issue as well?

**Comment by Yatekii** [comment]
Sorry for the not so nice words, but it's really to traefiks detriment that the docs are still wrong after 5 years and this is not fixed when it should be trivial.

**Comment by emilevauge** [comment]
@Yatekii you are absolutley true, the documentation needs more love. 
We are looking for some help, could you contribute on this? This would be great :)

**Comment by xsrf** [comment]
This is not a documentation issue. You shouldn't need to provide an irrelevant dummy port in the first place.
I think @rtribotte actually got it right.

**Comment by SR-G** [comment]
I fully agree. I lost HOURS trying to understand why Traefik was not working (as i was not aware of that dummy port being needed...) and i got the same answer : it's just a documentation issue, but it doesn't make sense (and i don't like having to configure "dummy things" in the tools i install, it's always confusing over time).

**Comment by emilevauge** [comment]
@xsrf my bad, I should have read more extensively. We will look into this.

**Comment by brunolnetto** [comment]
Guys, I read along, performed the actions you suggest (add the port label for load_balancer), deployed again the stack, registered the respective domain, but I am still unable to access the dashboard. Do you know where is my mistake? Tears have dropped more than once already.


```
services:
  traefik:
    image: traefik:v3.2.3
    hostname: '{{.Node.Hostname}}'
    command:
      - "--api.dashboard=true"
      - "--log.level=INFO"
      - "--providers.docker.endpoint=unix:///var/run/docker.sock"
      - "--providers.swarm.exposedByDefault=false"
      - "--providers.docker.network=traefik_network"
      - "--entrypoints.web.address=:80"
      - "--entrypoints.web.http.redirections.entryPoint.to=websecure"
      - "--entrypoints.web.http.redirections.entryPoint.scheme=https"
      - "--entrypoints.web.http.redirections.entrypoint.permanent=true"
      - "--entrypoints.websecure.address=:443"
      - "--entrypoints.web.transport.respondingTimeouts.idleTimeout=3600"
      - "--entrypoints.websecure.asDefault=true"
      - "--certificatesresolvers.letsencryptresolver.acme.httpchallenge=true"
      - "--certificatesresolvers.letsencryptresolver.acme.httpchallenge.entrypoint=web"
      - "--certificatesresolvers.letsencryptresolver.acme.storage=/etc/traefik/letsencrypt/acme.json"
      - "--certificatesresolvers.letsencryptresolver.acme.email=me@mail.com"
      - "--log.format=common"
      - "--log.filePath=/var/log/traefik/traefik.log"
      - "--accesslog=true"
      - "--accesslog.filepath=/var/log/traefik/access-log"

    deploy:
      mode: replicated
      replicas: 1
      placement:
        constraints:
          - node.role==manager

    labels:
      - "traefik.enable=true"
      - "traefik.web.routers.mydashboard.rule=Host(`traefik.example.com`)"
      - "traefik.web.routers.mydashboard.service=api@internal"
      - "traefik.web.routers.mydashboard.middlewares=myauth"
      - "traefik.web.services.mydashboard.loadbalancer.server.port=1337"
      - "traefik.web.middlewares.myauth.basicauth.users=test:$$secret_pwd_shh"

    volumes:
      - "vol_certificates:/etc/traefik/letsencrypt"
      - "/var/run/docker.sock:/var/run/docker.sock:ro"

    networks:
      - traefik_network

    ports:
      - target: 80
        published: 80
        mode: host
      - target: 443
        published: 443
        mode: host

volumes:
  vol_certificates:
    external: true
    name: volume_certificates

networks:
  traefik_network:
    attachable: true
    name: traefik_network
```

**Comment by pedroetb** [comment]
@brunolnetto after a quick view, is maybe because you are missing the entrypoint config label for your router (set to `http` or `websecure`).

**Comment by brunolnetto** [comment]
@pedroetb Thanks for quick answer. I editted my post to show my updated compose file. I am frustrated about not accomplishing the dashboard access. The authentication happened once, I do not know the reason it is not appearing anymore.

**Comment by pedroetb** [comment]
You are setting "traefik.http.routers.traefik.entrypoints=websecure" but
you need "traefik.http.routers.dashboard.entrypoints=websecure".

**Comment by danmuret** [comment]
@brunolnetto Did adding "traefik.http.routers.dashboard.entrypoints=websecure" resolve this?

**Comment by brunolnetto** [comment]
Yes, I added as following, with a valid subdomain-domain. It worked! 

```
    command:
      - "--api.dashboard=true"
      . . .

      labels:
        - "traefik.enable=true"
        - "traefik.http.routers.dashboard.rule=Host(`traefik.example.com`)"
        - "traefik.http.routers.dashboard.entrypoints=websecure"
        - "traefik.http.routers.dashboard.service=api@internal"
        - "traefik.http.routers.dashboard.tls.certresolver=letsencryptresolver"
        - "traefik.http.services.dummy-svc.loadbalancer.server.port=9999"
        - "traefik.http.routers.dashboard.middlewares=myauth"
        - "traefik.http.middlewares.myauth.basicauth.users=user:$apr1$hL/nwG5.$AHZhdlyxb6vVPOUBXu0Am/" # htpasswd -nbm user password
```

**Comment by nmengin** [comment]
Hey @brunolnetto, @pedroetb, @danmuret,

I'm glad to know that the configuration is working as expected now.
However, the conversation seems out of topic and I will hide it.

If you face another problem, please open a new issue or topic on the [community forum](https://community.traefik.io/).
Thanks in advance.
